src/ray/core_worker/transport/direct_task_transport.cc

```diff
@@ -115,24 +117,79 @@ void CoreWorkerDirectTaskSubmitter::OnWorkerIdle(const WorkerAddress &addr,
-  if (!queued_tasks_.empty()) {
-    RequestNewWorkerIfNeeded(queued_tasks_.front());
-  }
+  RequestNewWorkerIfNeeded(queued_tasks_.front());
 }
 
 void CoreWorkerDirectTaskSubmitter::RequestNewWorkerIfNeeded(
-    const TaskSpecification &resource_spec) {
+    const TaskSpecification &resource_spec, const rpc::Address *address) {
   if (worker_request_pending_) {
     return;
   }
-  RAY_CHECK_OK(lease_client_.RequestWorkerLease(resource_spec));
+  if (queued_tasks_.empty()) {
+    // We don't have any tasks to run, so no need to request a worker.
+    return;
+  }
+
+  std::shared_ptr<WorkerLeaseInterface> lease_client;
+  if (address && address->raylet_id() != "") {
+    // Connect to raylet.
+    ClientID raylet_id = ClientID::FromBinary(address->raylet_id());
+    auto it = remote_lease_clients_.find(raylet_id);
+    if (it == remote_lease_clients_.end()) {
+      RAY_LOG(DEBUG) << "Connecting to raylet " << raylet_id;
+      it =
+          remote_lease_clients_.emplace(raylet_id, lease_client_factory_(*address)).first;
+    }
+    RAY_LOG(DEBUG) << "Sending " << resource_spec.TaskId() << " to raylet " << raylet_id;
+    lease_client = it->second;
+  } else {
+    lease_client = local_lease_client_;
+  }
+
+  // NOTE(swang): We must copy the resource spec here because the resource spec
+  // may get swapped out by the time the callback fires.
+  TaskSpecification resource_spec_copy(resource_spec.GetMessage());
+  RAY_CHECK_OK(lease_client->RequestWorkerLease(
+      resource_spec_copy,
+      [this, resource_spec_copy, lease_client](
+          const Status &status, const rpc::WorkerLeaseReply &reply) mutable {
+        if (status.ok()) {
+          if (reply.retry_at_raylet_id() == "") {
+            RAY_LOG(DEBUG) << "Lease granted " << resource_spec_copy.TaskId();
+            HandleWorkerLeaseGranted({reply.address(), reply.port()},
+                                     std::move(lease_client));
+          } else {
+            absl::MutexLock lock(&mu_);
+            worker_request_pending_ = false;
+            rpc::Address address;
+            address.set_ip_address(reply.address());
+            address.set_port(reply.port());
+            address.set_raylet_id(reply.retry_at_raylet_id());
+            RequestNewWorkerIfNeeded(resource_spec_copy, &address);
+          }
+        } else {
+          RAY_LOG(DEBUG) << "Retrying lease request " << resource_spec_copy.TaskId();
+          absl::MutexLock lock(&mu_);
+          worker_request_pending_ = false;
+          if (lease_client != local_lease_client_) {
+            // A remote request failed. Retry the worker lease request locally
+            // if it's still in the queue.
+            // TODO(swang): Fail after some number of retries?
+            RAY_LOG(ERROR) << "Attempt to schedule task to remote node failed, retrying";
+            RequestNewWorkerIfNeeded(resource_spec_copy);
+          } else {
+            RAY_LOG(FATAL) << "Lost connection with local raylet.";
```

Comment:
Can you include the returned gRPC status in both these logs?

Reply:
Fixed.